net/net_evbuffer.c

```diff
@@ -3036,35 +3041,39 @@ int write_list_evbuffer(host_node_type *host_node_ptr, int type,
     int nodrop = flags & WRITE_MSG_NOLIMIT;
     int nodelay = flags & WRITE_MSG_NODELAY;
     struct event_info *e = host_node_ptr->event_info;
-    struct evbuffer *buf = evbuffer_new();
-    if (buf == NULL) {
-        rc = -1;
-        goto out;
-    }
-    if (evbuffer_add(buf, e->wirehdr[type], e->wirehdr_len) != 0) {
-        rc = -1;
-        goto out;
-    }
-    for (int i = 0; i < n; ++i) {
-        if (evbuffer_add(buf, iov[i].iov_base, iov[i].iov_len) != 0) {
-            rc = -1;
-            goto out;
-        }
-    }
+    size_t sz;
+    struct iovec v[1];
+    uint8_t *b;
+
     Pthread_mutex_lock(&e->wr_lk);
     if (!e->flush_buf) {
        rc = -3;
     } else if ((rc = skip_send(e, nodrop, 0)) == 0) {
-        if ((rc = evbuffer_add_buffer(e->flush_buf, buf)) == 0) {
+        sz = e->wirehdr_len;
+        for (int i = 0; i < n; ++i)
+            sz += iov[i].iov_len;
+
+        const int nv = evbuffer_reserve_space(e->flush_buf, sz, v, 1);
```

Comment:
Idea here was to do malloc without holding wr_lk

Reply:
Gotcha. I think I can malloc, memcpy outside `wr_lk`, and do an `add_reference()` while holding `wr_lk`. Would that be better?

Reply:
add_reference is additional malloc/free calls and again under lock. I'm not sure why we are touching write_list tbh.